File: app/src/main/java/com/example/cinemates/adapter/ItemsRecyclerViewAdapter.java
package com.example.cinemates.adapter;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.example.cinemates.databinding.ListItemMovieLongBinding;
import com.example.cinemates.databinding.ListItemMovieSmallBinding;
import com.example.cinemates.databinding.ListItemPersonLongBinding;
import com.example.cinemates.databinding.ListItemPersonSmallBinding;
import com.example.cinemates.model.data.Cast;
import com.example.cinemates.model.data.Movie;
import com.example.cinemates.model.data.Person;
import com.example.cinemates.view.ui.ActorDetailsActivity;
import com.example.cinemates.view.ui.MovieDetailsActivity;
import com.example.cinemates.util.MyDiffUtilCallbacks;
import com.example.cinemates.util.RecyclerViewEmptySupport;
import com.example.cinemates.util.ViewSize;

import java.util.ArrayList;
import java.util.List;


/**
 * @author Antonio Di Nuzzo
 * Created 15/12/2021 at 16:36
 */
public class ItemsRecyclerViewAdapter<T> extends RecyclerViewEmptySupport.Adapter<RecyclerView.ViewHolder> {
    private final List<T> dataList = new ArrayList<>();
    private final int PERSON = 0, MOVIE = 1;
    private final ViewSize mViewSize;

    public ItemsRecyclerViewAdapter(ViewSize viewSize) {
        mViewSize = viewSize;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        RecyclerView.ViewHolder viewHolder;

        switch (viewType) {
            case MOVIE:
                if (mViewSize == ViewSize.SMALL) {
                    ListItemMovieSmallBinding smallMovieBinding = ListItemMovieSmallBinding.inflate(layoutInflater, parent, false);
                    viewHolder = new MovieViewHolder(smallMovieBinding);
                } else {
                    ListItemMovieLongBinding longMovieBinding = ListItemMovieLongBinding.inflate(layoutInflater, parent, false);
                    viewHolder = new MovieViewHolder(longMovieBinding);
                }

                break;
            case PERSON:
                if (mViewSize == ViewSize.SMALL) {
                    ListItemPersonSmallBinding smallPersonBinding = ListItemPersonSmallBinding.inflate(layoutInflater, parent, false);
                    viewHolder = new PersonViewHolder(smallPersonBinding);
                } else {
                    ListItemPersonLongBinding defaultBinding = ListItemPersonLongBinding.inflate(layoutInflater, parent, false);
                    viewHolder = new PersonViewHolder(defaultBinding);
                }

                break;

            default:
                throw new IllegalStateException("Unexpected value: " + viewType);
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case MOVIE:
                Movie movie = (Movie) dataList.get(position);
                switch (mViewSize) {

                    case LONG:
                        ((MovieViewHolder) holder).mMovieLongBinding.setMovie(movie);
                        ((MovieViewHolder) holder).mMovieLongBinding.executePendingBindings();
                        ((MovieViewHolder) holder).mMovieLongBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {

                                Intent intent = new Intent(view.getContext(), MovieDetailsActivity.class);
                                intent.putExtra("movie", movie);
                                view.getContext().startActivity(intent);

                            }
                        });
                        break;
                    case SMALL:
                        ((MovieViewHolder) holder).mMovieSmallBinding.setMovie(movie);
                        ((MovieViewHolder) holder).mMovieSmallBinding.executePendingBindings();
                        ((MovieViewHolder) holder).mMovieSmallBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {

                                Intent intent = new Intent(view.getContext(), MovieDetailsActivity.class);
                                intent.putExtra("movie", movie);
                                view.getContext().startActivity(intent);

                            }
                        });
                        break;
                }

                break;
            case PERSON:
                switch (mViewSize) {
                    case LONG:
                        ((PersonViewHolder) holder).mLongBinding.setActor((Cast) dataList.get(position));
                        ((PersonViewHolder) holder).mLongBinding.executePendingBindings();
                        ((PersonViewHolder) holder).mLongBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                Intent intent = new Intent(view.getContext(), ActorDetailsActivity.class);
                                intent.putExtra("person", (Person) dataList.get(position));
                                view.getContext().startActivity(intent);
                            }
                        });
                        break;
                    case SMALL:
                        ((PersonViewHolder) holder).mSmallBinding.setPerson((Person) dataList.get(position));
                        ((PersonViewHolder) holder).mSmallBinding.executePendingBindings();
                        ((PersonViewHolder) holder).mSmallBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                Intent intent = new Intent(view.getContext(), ActorDetailsActivity.class);
                                intent.putExtra("person", (Person) dataList.get(position));
                                view.getContext().startActivity(intent);
                            }
                        });
                        break;
                }

                break;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull List<Object> payloads) {
        if (payloads.isEmpty())
            super.onBindViewHolder(holder, position, payloads);
        else {
            switch (holder.getItemViewType()) {
                case MOVIE:
                    Movie movie = (Movie) dataList.get(position);
                    switch (mViewSize) {

                        case LONG:
                            ((MovieViewHolder) holder).mMovieLongBinding.setMovie(movie);
                            ((MovieViewHolder) holder).mMovieLongBinding.executePendingBindings();
                            ((MovieViewHolder) holder).mMovieLongBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {

                                    Intent intent = new Intent(view.getContext(), MovieDetailsActivity.class);
                                    intent.putExtra("movie", movie);
                                    view.getContext().startActivity(intent);

                                }
                            });
                            break;
                        case SMALL:
                            ((MovieViewHolder) holder).mMovieSmallBinding.setMovie(movie);
                            ((MovieViewHolder) holder).mMovieSmallBinding.executePendingBindings();
                            ((MovieViewHolder) holder).mMovieSmallBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {

                                    Intent intent = new Intent(view.getContext(), MovieDetailsActivity.class);
                                    intent.putExtra("movie", movie);
                                    view.getContext().startActivity(intent);

                                }
                            });
                            break;
                    }
                    break;
                case PERSON:
                    switch (mViewSize) {
                        case LONG:
                            ((PersonViewHolder) holder).mLongBinding.setActor((Cast) dataList.get(position));
                            ((PersonViewHolder) holder).mLongBinding.executePendingBindings();
                            ((PersonViewHolder) holder).mLongBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {
                                    Intent intent = new Intent(view.getContext(), ActorDetailsActivity.class);
                                    intent.putExtra("person", (Person) dataList.get(position));
                                    view.getContext().startActivity(intent);
                                }
                            });
                            break;
                        case SMALL:
                            ((PersonViewHolder) holder).mSmallBinding.setPerson((Person) dataList.get(position));
                            ((PersonViewHolder) holder).mSmallBinding.executePendingBindings();
                            ((PersonViewHolder) holder).mSmallBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {
                                    Intent intent = new Intent(view.getContext(), ActorDetailsActivity.class);
                                    intent.putExtra("person", (Person) dataList.get(position));
                                    view.getContext().startActivity(intent);
                                }
                            });
                            break;
                    }
                    break;
            }
        }
    }

    //Returns the view type of the item at position for the purposes of view recycling.
    @Override
    public int getItemViewType(int position) {
        if (dataList.get(position) instanceof Movie) {
            return MOVIE;
        } else if ((dataList.get(position) instanceof Person)) {
            return PERSON;
        }
        return -1;
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    public void addItems(List<T> dataList) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new MyDiffUtilCallbacks(this.dataList, dataList));
        diffResult.dispatchUpdatesTo(this);

        this.dataList.clear();
        this.dataList.addAll(dataList);
        notifyDataSetChanged();
    }


}

File: app/src/main/java/com/example/cinemates/adapter/ItemsRecyclerViewAdapter.kt
package com.example.cinemates.adapter

import android.view.View
import android.view.ViewGroup
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.example.cinemates.NavGraphDirections
import com.example.cinemates.databinding.ListItemMovieLongBinding
import com.example.cinemates.databinding.ListItemMovieSmallBinding
import com.example.cinemates.databinding.ListItemPersonLongBinding
import com.example.cinemates.databinding.ListItemPersonSmallBinding
import com.example.cinemates.model.data.Cast
import com.example.cinemates.model.data.Movie
import com.example.cinemates.model.data.Person
import com.example.cinemates.util.ViewSize
import com.example.cinemates.util.inflater


/**
 * @author Antonio Di Nuzzo
 * Created 15/12/2021 at 16:36
 */
class ItemsRecyclerViewAdapter<T>(private val viewType: ViewSize) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var dataList: MutableList<T> = arrayListOf()

    private companion object {
        const val PERSON = 0
        const val MOVIE = 1

        class MovieSmallViewHolder(val binding: ListItemMovieSmallBinding) :
            RecyclerView.ViewHolder(binding.root) {
        }

        class MovieLongViewHolder(val binding: ListItemMovieLongBinding) :
            RecyclerView.ViewHolder(binding.root) {
        }

        class PersonSmallViewHolder(val binding: ListItemPersonSmallBinding) :
            RecyclerView.ViewHolder(binding.root) {
        }

        class PersonLongViewHolder(val binding: ListItemPersonLongBinding) :
            RecyclerView.ViewHolder(binding.root) {
        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val viewHolder: RecyclerView.ViewHolder = when (viewType) {
            MOVIE -> if (this.viewType == ViewSize.SMALL) {
                MovieSmallViewHolder(parent inflater ListItemMovieSmallBinding::inflate)
            } else {
                MovieLongViewHolder(parent inflater ListItemMovieLongBinding::inflate)
            }
            PERSON -> if (this.viewType == ViewSize.SMALL) {
                PersonSmallViewHolder(parent inflater ListItemPersonSmallBinding::inflate)
            } else {
                PersonLongViewHolder(parent inflater ListItemPersonLongBinding::inflate)
            }
            else -> throw IllegalStateException("Unexpected value: $viewType")
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            MOVIE -> {
                val movie = dataList[position] as Movie
                when (viewType) {
                    ViewSize.LONG -> {
                        (holder as MovieLongViewHolder).binding.apply {
                            this.movie = movie
                            executePendingBindings()
                            root.setOnClickListener { view -> navigateToMovieDetails(view, movie) }
                        }
                    }
                    ViewSize.SMALL -> {
                        (holder as MovieSmallViewHolder).binding.apply {
                            this.movie = movie
                            executePendingBindings()
                            root.setOnClickListener { view ->
                                navigateToMovieDetails(
                                    view,
                                    movie
                                )
                            }
                        }
                    }
                }
            }
            PERSON -> {
                val person = dataList[position] as Person
                when (viewType) {
                    ViewSize.LONG -> {
                        (holder as PersonLongViewHolder).binding.apply {
                            actor = person as Cast
                            executePendingBindings()
                            root.setOnClickListener { view ->
                                navigateToActorDetails(view, person)
                            }
                        }
                    }
                    ViewSize.SMALL -> {
                        (holder as PersonSmallViewHolder).binding.apply {
                            this.person = person
                            executePendingBindings()
                            root.setOnClickListener { view ->
                                navigateToActorDetails(view, person)
                            }
                        }
                    }
                }
            }
        }
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int,
        payloads: List<Any>
    ) {
        if (payloads.isEmpty()) super.onBindViewHolder(holder, position, payloads) else {
            when (holder.itemViewType) {
                MOVIE -> {
                    val movie = dataList[position] as Movie
                    when (viewType) {
                        ViewSize.LONG -> {
                            (holder as MovieLongViewHolder).binding.apply {
                                this.movie = movie
                                executePendingBindings()
                                root.setOnClickListener { view ->
                                    navigateToMovieDetails(
                                        view,
                                        movie
                                    )
                                }
                            }
                        }
                        ViewSize.SMALL -> {
                            (holder as MovieSmallViewHolder).binding.apply {
                                this.movie = movie
                                executePendingBindings()
                                root.setOnClickListener { view ->
                                    navigateToMovieDetails(
                                        view,
                                        movie
                                    )
                                }
                            }
                        }
                    }
                }
                PERSON -> {
                    val person = dataList[position] as Person
                    when (viewType) {
                        ViewSize.LONG -> {
                            (holder as PersonLongViewHolder).binding.apply {
                                actor = person as Cast
                                executePendingBindings()
                                root.setOnClickListener { view ->
                                    navigateToActorDetails(view, person)
                                }
                            }
                        }
                        ViewSize.SMALL -> {
                            (holder as PersonSmallViewHolder).binding.apply {
                                this.person = person
                                executePendingBindings()
                                root.setOnClickListener { view ->
                                    navigateToActorDetails(view, person)
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun navigateToActorDetails(view: View, person: Person) {
        val action = NavGraphDirections.actionGlobalActorDetailsFragment(person)
        findNavController(view).navigate(action)
    }

    private fun navigateToMovieDetails(view: View, movie: Movie) {
        val action = NavGraphDirections.actionGlobalMovieDetailsFragment(movie)
        findNavController(view).navigate(action)
    }

    //Returns the view type of the item at position for the purposes of view recycling.
    override fun getItemViewType(position: Int): Int {
        val value = dataList[position]
        if (value is Movie) {
            return MOVIE
        } else if (value is Person) {
            return PERSON
        }
        return -1
    }

    override fun getItemCount(): Int {
        return dataList.size
    }

    fun addItems(dataList: MutableList<T>) {
        this.dataList.clear()
        this.dataList.addAll(dataList)
        notifyDataSetChanged()
    }


}



